src/utils/sys_fill_exact.rs

```diff
@@ -0,0 +1,36 @@
+use crate::Error;
+use core::mem::MaybeUninit;
+
+mod get_errno;
+mod sanitizer;
+
+pub(crate) use get_errno::get_errno;
+
+pub(crate) fn sys_fill_exact(
+    mut buf: &mut [MaybeUninit<u8>],
+    sys_fill: impl Fn(&mut [MaybeUninit<u8>]) -> libc::ssize_t,
+) -> Result<(), Error> {
+    while !buf.is_empty() {
+        let res = sys_fill(buf);
+        match res {
+            res if res > 0 => {
+                let len = usize::try_from(res).map_err(|_| Error::UNEXPECTED)?;
+                let (l, r) = buf.split_at_mut_checked(len).ok_or(Error::UNEXPECTED)?;
+                unsafe { sanitizer::unpoison(l) };
```

Comment:
Note that it applies `unpoison` to all users of `sys_fill_exact`. It may be redundant, but AFAIK it has no negative consequences (outside of outlandish scenarios of misbehaving `libc`), while allowing to simplify the code a fair bit.